source/slang/slang-language-server-semantic-tokens.cpp

```diff
@@ -0,0 +1,611 @@
+             .pathInfo.foundPath.getUnownedSlice()
+             .endsWithCaseInsensitive(fileName))
+        return;
+
+    maybeDispatchCallback(decl);
+    if (auto funcDecl = as<FunctionDeclBase>(decl))
+    {
+        visitStmt(funcDecl->body);
+        visitExpr(funcDecl->returnType.exp);
+    }
+    else if (auto propertyDecl = as<PropertyDecl>(decl))
+    {
+        visitExpr(propertyDecl->type.exp);
+    }
+    else if (auto varDecl = as<VarDeclBase>(decl))
+    {
+        visitExpr(varDecl->type.exp);
+        visitExpr(varDecl->initExpr);
+    }
+    else if (auto genericDecl = as<GenericDecl>(decl))
+    {
+        visitDecl(genericDecl->inner);
+    }
+    else if (auto typeConstraint = as<TypeConstraintDecl>(decl))
+    {
+        visitExpr(typeConstraint->getSup().exp);
+    }
+    else if (auto typedefDecl = as<TypeDefDecl>(decl))
+    {
+        visitExpr(typedefDecl->type.exp);
+    }
+    if (auto container = as<ContainerDecl>(decl))
+    {
+        for (auto member : container->members)
+        {
+            visitDecl(member);
+        }
+    }
+}
+template <typename CallbackFunc>
+void ASTIterator<CallbackFunc>::visitExpr(Expr* expr)
+{
+    ASTIteratorExprVisitor visitor(this);
+    visitor.dispatchIfNotNull(expr);
+}
+template <typename CallbackFunc>
+void ASTIterator<CallbackFunc>::visitStmt(Stmt* stmt)
+{
+    ASTIteratorStmtVisitor visitor(this);
+    visitor.dispatchIfNotNull(stmt);
+}
+
+template <typename Func>
+void iterateAST(UnownedStringSlice fileName, SourceManager* manager, SyntaxNode* node, const Func& f)
+{
+    ASTIterator<Func> iter(f, manager, fileName);
+    if (auto decl = as<Decl>(node))
+    {
+        iter.visitDecl(decl);
+    }
+    else if (auto expr = as<Expr>(node))
+    {
+        iter.visitExpr(expr);
+    }
+    else if (auto stmt = as<Stmt>(node))
+    {
+        iter.visitStmt(stmt);
+    }
+}
+
+const char* kSemanticTokenTypes[] = {
+    "type", "enumMember", "variable", "parameter", "function", "property", "namespace"};
+
+static_assert(SLANG_COUNT_OF(kSemanticTokenTypes) == (int)SemanticTokenType::_NormalText, "kSemanticTokenTypes must match SemanticTokenType");
+
+SemanticToken _createSemanticToken(SourceManager* manager, SourceLoc loc, Name* name)
+{
+    SemanticToken token;
+    auto humaneLoc = manager->getHumaneLoc(loc, SourceLocType::Actual);
+    token.line = (int)(humaneLoc.line - 1);
+    token.col = (int)(humaneLoc.column - 1);
+    token.length =
+        name ? (int)(name->text.getLength()) : 0;
+    token.type = SemanticTokenType::_NormalText;
+    return token;
+}
+
+List<SemanticToken> getSemanticTokens(Linkage* linkage, Module* module, UnownedStringSlice fileName)
+{
+    auto manager = linkage->getSourceManager();
+
+    List<SemanticToken> result;
+    auto maybeInsertToken = [&](const SemanticToken& token)
+    {
+        if (token.line >= 0 && token.col >= 0 && token.length > 0 &&
+            token.type != SemanticTokenType::_NormalText)
```

Comment:
I know it's elsewhere in our code but any identifier prefixed with _C where C is a capital letter is compiler reserved.

Reply:
Fixed.